### ISSUE TITLE
file_path_key fails in active_support/cache/file_store.rb for long paths

### ISSUE THREAD

**Issue by madej** [rails/rails#21576]
This issue leads to exception in URI.decode_www_form_component if cutting path was in a middle of separator (%2F) or just returns non existing key.

in def file_path_key(path)
fname = path[cache_path.to_s.size..-1].split(File::SEPARATOR, 4).last

doesn't take on account path might be cut into pieces of FILENAME_MAX_SIZE in key_file_path.
changing it to:
fname = path[cache_path.to_s.size..-1].split(File::SEPARATOR, 4).last.gsub(File::SEPARATOR, '')

fixes the problem.

**Comment by agis-** [comment]
@madej What rails version are you on? Also, can you provide a script that reproduces the issue?

**Comment by madej** [comment]
@agis- 4.2.4, but i can see the problem still exists on master. To reproduce it you need to create a key which is longer then FILENAME_MAX_SIZE, but shorter than FILEPATH_MAX_SIZE. In real life it might be a result array of active_record query. Just set caching to file_store and run it in rails console.

```ruby
begin
  key = ['lol', 'lo'] * 120
  Rails.cache.fetch key, expires_in: 1.seconds do 'lol' end
  sleep 2
  Rails.cache.cleanup
end
```

**Comment by madej** [comment]
Just one remark. The solution i proposed just fixes this problem, but to me honest i think better solution might be removing cutting key into pieces and always counting md5. This way we could remove URI encoding and decoding, thus get some performance improvement.

**Comment by madej** [comment]
In rails 5 problem still exists. Though, to reproduce it one must use smaller keys, so in example above use:
`key = ['lol', 'lo'] * 80`

**Comment by madej** [comment]
I've just noticed that cleanup doesn't work at all on master.
```ruby
def cleanup(options = nil)
  options = merged_options(options)
  search_dir(cache_path) do |fname|
    key = file_path_key(fname)
    entry = read_entry(key, options)
    delete_entry(key, options) if entry && entry.expired?
  end
end
```

To `read_entry` and `delete_entry` is used key whilst these functions work on file path.

I believe this function should look more like:
```ruby
def cleanup(options = nil)
  options = merged_options(options)
  search_dir(cache_path) do |fname|
    entry = read_entry(fname, options)
    delete_entry(fname, options) if entry && entry.expired?
  end
end
```

After this change only `delete_matched` will use messed up `file_path_key`.

**Comment by maxcodes** [comment]
According to https://github.com/rails/rails/issues/30788#issuecomment-400509777, this issue is already fixed on v5.2.0.beta1 and should be closed 🙂
